tempesta_fw/http_parser.c

```diff
@@ -103,17 +92,105 @@ memchreol(const unsigned char *s, size_t n)
+		tfw_http_msg_hdr_chunk_fixup(msg, data,			\
+					     __data_offset(pos));	\
+		__data_move_to(pos);					\
+	}								\
+} while (0)
+#define __msg_hdr_finish(pos)						\
+({									\
+	int r = tfw_http_msg_hdr_close(msg, msg->parser._hdr_tag);	\
+	int eolen = 1 + !!((parser->_tmp.eol & 0xff) == 0xda);		\
+	tfw_str_set_eolen(&parser->hdr, eolen);				\
+	r;								\
+})
+
+/**
+ * The following set of macroses is intended to use for generic fields
+ * processing while parsing HTTP status-line. As with headers, @__msg_field_open
+ * macro is used for field openning, @__msg_field_fixup is used for updating,
+ * and @__msg_field_finish is used when field needs to be finished. The latter
+ * means that the underlying TfwStr flag TFW_STR_COMPLETE must be raised.
+ */
+#define __msg_field_open(field, pos)					\
+do {									\
+	tfw_http_msg_set_data(msg, field, pos);				\
+	__data_move_to(pos);						\
+} while (0)
+#define __msg_field_fixup(field, pos)					\
+do {									\
+	if (__data_offset(pos)) {					\
+		tfw_http_msg_field_chunk_fixup(msg, field, data,	\
+					       __data_offset(pos));	\
+		__data_move_to(pos);					\
+	}								\
+} while (0)
+#define __msg_field_finish(field, pos)					\
+do {									\
+	__msg_field_fixup(field, pos);					\
+	(field)->flags |= TFW_STR_COMPLETE;				\
+} while (0)
+
+/**
+ * Special field @crlf is used to track HTTP empty-line while handling requests
+ * and responses. It can be treated as generic @msg field of type TfwStr but
+ * there is no reason to do any fixup on it as it always has zero length and
+ * non-zero EOL that must be updated properly at field closing. So finishing
+ * macro needs to be specific.
+ */
+#define __msg_crlf_open(pos)						\
+	__msg_field_open(&msg->crlf, pos)
+#define __msg_crlf_finish(pos)						\
+do {									\
+	int eolen = 1 + !!((msg->parser._tmp.eol & 0xff) == 0xda);	\
+	tfw_str_set_eolen(&msg->crlf, eolen);				\
+	msg->crlf.flags |= TFW_STR_COMPLETE;				\
+} while (0)
+
+/**
  * GCC 4.8 (CentOS 7) does a poor work on memory reusage of automatic local
  * variables in nested blocks, so we declare all required temporal variables
  * used in the defines below here to reduce stack frame usage.
  * Since the variables are global now, be careful with them.
  */
-#define __FSM_DECLARE_VARS()						\
-int __fsm_const_state;							\
-/* Declare FSM automatic variables, the variables have only local sense. */ \
-int __fsm_n __attribute__((unused));					\
-size_t __fsm_sz __attribute__((unused));				\
-unsigned char *__fsm_ch __attribute__((unused));
+#define __FSM_DECLARE_VARS(ptr)						\
+	TfwHttpMsg	* msg = (TfwHttpMsg *)(ptr);			\
+	TfwHttpParser	* parser = &msg->parser;			\
+	unsigned char	* p = data;					\
+	unsigned char	c = *p;						\
+	int		__fsm_const_state;				\
+	int		__maybe_unused __fsm_n;				\
+	size_t		__maybe_unused __fsm_sz;			\
+	unsigned char	__maybe_unused * __fsm_ch;			\
+	TfwStr		__maybe_unused * chunk = &parser->_tmp_chunk;	\
+	;
```

Comment:
Actually compilers don't do the job well, e.g. https://github.com/natsys/tempesta/blob/master/tempesta_fw/http_parser.c#L106 , so we have used the rule as a coding style.

However, you're right - I just made a simple test:

        $ cat stack.c 
        extern int call(long a, long b, long c);
        
        int func (int argc, char *argv[])
        {
        	volatile char b = *argv[0];
        	volatile int a = argc;
        	volatile short e = *argv[1];
        	volatile long d = *(long *)argv[1];
        	return call(a + b, a + d, b - e);
        }
        $ gcc --version|head -1
        gcc (GCC) 5.1.1 20150618 (Red Hat 5.1.1-4)
        $ gcc -O2 -c -S -o stack.s stack.c
        $ grep -A 20 '\<func:' stack.s
        func:
        .LFB0:
        	.cfi_startproc
        	subq	$24, %rsp
        	.cfi_def_cfa_offset 32
        	movq	(%rsi), %rax
        	movzbl	(%rax), %eax
        	movb	%al, 1(%rsp)
        	movq	8(%rsi), %rax
        	movl	%edi, 4(%rsp)
        	movsbw	(%rax), %dx
        	movq	(%rax), %rax
        	movw	%dx, 2(%rsp)
        	movq	%rax, 8(%rsp)
        	movsbl	1(%rsp), %edx
        	movswl	2(%rsp), %r8d
        	movslq	4(%rsp), %rsi
        	movq	8(%rsp), %rcx
        	movsbl	1(%rsp), %edi
        	movl	4(%rsp), %eax
        	subl	%r8d, %edx

So compiler uses only 16 bytes for the data and properly pack them on stack.